### PR TITLE
Fix MSON enum bug

### LINKED ISSUE #95: MSON - enum bug - samples/default parsed as array
## Input
```
Example

## GET /
+ Attributes
    + id: 20 (enum[number], optional) - Database ID

        Additional description

        + Default: 1

        + Sample: 23

+ Response 201
```
## Result (partial):

```
... 
"value": {
          "element": "enum",
          "attributes": {
            "samples": [
              [
                "23"
              ]
            ],
            "default": [
              "1"
            ]
...
```
## BUG
- attributes.samples and attributes.default saved as array
- individual values are typed as "StringElement" - it should hold type specification from `enum[number]`

### PR DESCRIPTION
**Pull request by klokane**
This patch fix #95:
- MSON Enum attribudes samples && default - are now presented as values (instead of array)
- nested types (eg. enum[number]) are now propagated deeply into members, samples and default

There is a few things to discuss before **merge** 

missing: 
- propagate description of members into refract

### CODE REVIEW

File: src/RefractDataStructure.cc [hunk 1]
```diff
@@ -143,25 +146,35 @@ namespace drafter {
...
-    static refract::IElement* MsonElementToRefract(const mson::Element& mse);
+    static refract::IElement* MsonElementToRefract(const mson::Element& mse, mson::BaseTypeName defaultNestedType = mson::StringTypeName);
 
-    RefractElements MsonElementsToRefract(const mson::Elements& elements) {
+    RefractElements MsonElementsToRefract(const mson::Elements& elements, mson::BaseTypeName defaultNestedType = mson::StringTypeName) {
         RefractElements result;
-        std::transform(elements.begin(), elements.end(), std::back_inserter(result), MsonElementToRefract);
+        //std::transform(elements.begin(), elements.end(),
+        //               std::back_inserter(result),
+        //               std::bind2nd(std::ptr_fun(MsonElementToRefract), nestedTypeName));
```
Comment (pksunkara) [inline src/RefractDataStructure.cc]: Shouldn't we remove this?

**Comment by pksunkara** [comment]
I haven't reviewed this completely. But I don't think `defaultNestedType` will correctly parse the following example:

```
# User
- tag (enum)
    - pavan (string)
    - 4 (number)
```

File: src/RefractDataStructure.cc (hunk 1 above)
Comment (klokane) [inline src/RefractDataStructure.cc]: There is some trouble with `bind2nd` `ptr_fun` and `MsonElementToRefract`, I will check it later.
Temporary it is replaced by loop below.

**Comment by klokane** [comment]
Reason?

**Comment by pksunkara** [comment]
I haven't tried it. Can you add a test for the above example?

**Comment by klokane** [comment]
Sure, I can. But I mean that you have, at least, some idea why it should be parsed wrong...

**Comment by klokane** [comment]
ready for review

**Comment by klokane** [comment]
So pls, do not merge, will go thru examples and check them against to result of discussion

**Comment by klokane** [comment]
@pksunkara @z can you please look at following fixture
https://github.com/apiaryio/drafter/blob/shared/refract/test/fixtures/api/resource-parameters.apib
I mean values for "default"  and "samples" https://github.com/apiaryio/drafter/blob/shared/refract/test/fixtures/api/resource-parameters.json#L47
should be presented as "array", so
```apib
                     "attributes": {
                        "samples": [
                          23
                        ],
                        "default": 1
                      },
```

should be 
```apib
                     "attributes": {
                        "samples": [
                          [ 23 ]
                        ],
                        "default": [ 1 ]
                      },
```
according to our latest discussion.

But please be warn there are attributes part of  "API Description Namespace" not "Data Structure Namespace" but

**Comment by klokane** [comment]
I check all enum|array in "mson" fixrtures it looks correct

File: test/fixtures/mson/enum-sample.json [hunk 2]
```diff
@@ -27,6 +27,28 @@
...
+                  "content": {
+                    "key": {
+                      "element": "string",
+                      "content": "e2"
+                    },
+                    "value": {
+                      "element": "enum",
                       "attributes": {
                         "samples": [
                           [
```
Comment (pksunkara) [inline test/fixtures/mson/enum-sample.json]: This should be

```json
[
  {
    "element": "string",
     "content": "3"
  },
  {
    "element": "string",
     "content": "4"
  }
]
```

Comment (pksunkara) [inline test/fixtures/mson/enum-sample.json]: Similar problem in e4.

Comment (klokane) [inline test/fixtures/mson/enum-sample.json]: Yes, but firstly must be fixed #100 :)

File: test/fixtures/mson/enum-sample.json [hunk 3]
```diff
@@ -82,6 +107,90 @@
...
+                      "element": "enum",
+                      "attributes": {
+                        "samples": [
+                          [
+                            3,
+                            4
+                          ],
+                          [
+                            5,
+                            6
```
Comment (pksunkara) [inline test/fixtures/mson/enum-sample.json]: Similar problem as [above](https://github.com/apiaryio/drafter/pull/94/files#r39053375)

**Comment by pksunkara** [comment]
I am going to change the parameters in a new PR.

**Comment by klokane** [comment]
ok, thank for clear

**Comment by pksunkara** [comment]
This is not compiling in clang. Please make a separate PR to fix that. I am merging this for now.

**Pull request merged**